File: backend/src/main/java/ch/noseryoung/domain/recur/models/task/Task.java
package ch.noseryoung.domain.recur.models.task;

import ch.noseryoung.domain.recur.auth.model.User;
import ch.noseryoung.domain.recur.models.group.Project;
import ch.noseryoung.domain.recur.models.notification.NotificationLog;

import java.time.Instant;
import java.time.LocalDate;
import java.util.*;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import com.fasterxml.jackson.annotation.JsonIgnore;

import ch.noseryoung.domain.recur.enums.task.Category;
import ch.noseryoung.domain.recur.enums.task.Frequency;
import ch.noseryoung.domain.recur.enums.notification.ReminderLeadTime;
import lombok.*;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;

// toBuilder=true wird gebraucht, um für die Response eine transiente Kopie
// mit maskierten Mitgliedern (siehe TaskService#maskMembers) zu bauen, ohne
// die verwaltete Entity selbst zu verändern (kein versehentliches Zurück-
// schreiben anonymisierter Namen in die echten Zuweisungs-/Archiv-Tabellen).
@Builder(toBuilder = true)
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Entity
@Table(name = "task")
public class Task {

        // Marker-Interface für Validation-Groups: Constraints unten gelten nur beim
        // Erstellen (POST), nicht bei partiellen PATCH-Updates, da patchTask()
        // bewusst leere Felder als "nicht ändern" interpretiert.
        public interface OnCreate {
        }

        @Id
        @GeneratedValue(strategy = GenerationType.UUID)
        @Column(name = "id")
        private UUID id;

        @NotBlank(message = "Name ist erforderlich", groups = OnCreate.class)
        @Size(min = 2, max = 40, message = "Name muss zwischen 2 und 40 Zeichen lang sein", groups = OnCreate.class)
        @Column(name = "name")
        private String name;

        @NotNull(message = "Kategorie ist erforderlich", groups = OnCreate.class)
        @Enumerated(EnumType.STRING)
        @Column(name = "category")
        private Category category;

        @NotNull(message = "Frequenz ist erforderlich", groups = OnCreate.class)
        @Enumerated(EnumType.STRING)
        @Column(name = "frequency")
        private Frequency frequency;

        @NotBlank(message = "Beschreibung ist erforderlich", groups = OnCreate.class)
        @Size(max = 200, message = "Beschreibung darf maximal 200 Zeichen lang sein", groups = OnCreate.class)
        @Column(name = "description")
        private String description;

        @NotNull(message = "Fälligkeitsdatum ist erforderlich", groups = OnCreate.class)
        @Column(name = "date_until")
        private Instant dateUntil;

        @Column(name = "progress")
        private Double progress;

        @CreationTimestamp
        @Column(name = "date_created", updatable = false)
        private Instant dateCreated;

        // Wird bei jedem Save automatisch neu gesetzt (Hibernate). Dient dem
        // Frontend-Polling als Last-Write-Wins-Vergleichswert, damit ein Sync-Poll
        // laufende optimistische UI-Updates nicht überschreibt.
        @UpdateTimestamp
        @Column(name = "updated_at")
        private Instant updatedAt;

        @Column(name = "days_in_span")
        private Integer daysInSpan;

        @Column(name = "amount_did")
        private Integer amountDid;

        // Zeitpunkt des letzten Fortschritts-Increments (Abhaken-Button /
        // "Als erledigt markieren") - Basis für die Rolling-Window-Sperre des
        // Abhaken-Buttons pro Frequenz-Intervall (#136). Wird bei resetProgress
        // wieder auf null gesetzt.
        @Column(name = "last_amount_did_at")
        private Instant lastAmountDidAt;

        // Completion-Historie pro Kalendertag (#152) - Source of Truth für
        // amountDid/progress/lastAmountDidAt, die daraus abgeleitet werden (siehe
        // TaskUtil#syncCompletions). Nur bei persönlichen Tasks genutzt (owner !=
        // null), Projekt-Tasks bleiben bei der bisherigen completedBy-Logik.
        @Builder.Default
        @ElementCollection
        @CollectionTable(name = "task_completion", joinColumns = @JoinColumn(name = "task_id"), uniqueConstraints = @UniqueConstraint(columnNames = {
                        "task_id", "completion_date" }))
        @Column(name = "completion_date")
        private Set<LocalDate> completions = new HashSet<>();

        @Builder.Default
        @Column(name = "is_favorite", nullable = false, columnDefinition = "boolean default false")
        private Boolean isFavorite = false;

        @Builder.Default
        @Column(name = "is_archived", nullable = false, columnDefinition = "boolean default false")
        private Boolean isArchived = false;

        // Bei einem persönlichen Task gesetzt. Ist der Task stattdessen einem
        // Projekt zugeordnet (siehe unten), ist er ein geteiltes Item ohne
        // einzelnen Besitzer - dann bleibt owner null und der Zugriff läuft
        // über die Mitgliedschaft in project.group.
        @ManyToOne(fetch = FetchType.LAZY)
        @JoinColumn(name = "user_id")
        private User owner;

        @Column(name = "duration_minutes")
        private Integer durationMinutes;

        @Column(name = "start_time")
        private Instant startTime;

        // Transient, nicht die persistierte Task-Zeile: der Erinnerungs-Vorlauf
        // ist ein Override pro (Task, Empfänger) - siehe TaskReminderOverride -
        // nicht ein einzelner Wert für den ganzen (bei Projekt-Tasks geteilten)
        // Task. Wird von TaskService#maskMembers für den jeweiligen Betrachter
        // befüllt, bevor der Task in einer Response landet.
        @Transient
        private ReminderLeadTime reminderLeadTime;

        // Optionale Zuordnung zu einem Gruppen-Projekt. Gesetzt <=> der Task ist
        // ein geteiltes Checklisten-Item für alle Mitglieder der Projekt-Gruppe
        // statt eines persönlichen Tasks.
        @ManyToOne(fetch = FetchType.LAZY)
        @JoinColumn(name = "project_id")
        private Project project;

        // Wird gesetzt, sobald der Task (für ein Projekt) als erledigt markiert
        // wird, und wieder auf null gesetzt sobald er zurückgesetzt wird. Nur für
        // geteilte Projekt-Tasks relevant, bei persönlichen Tasks ungenutzt.
        @ManyToOne(fetch = FetchType.LAZY)
        @JoinColumn(name = "completed_by")
        private User completedBy;

        // Self-Service: Gruppenmitglieder können sich selbst einem geteilten
        // Projekt-Task zuweisen/abmelden (TaskService#assignSelf/unassignSelf).
        // Nur für Projekt-Tasks relevant.
        @Builder.Default
        @ManyToMany(fetch = FetchType.LAZY)
        @JoinTable(name = "task_assignee", joinColumns = @JoinColumn(name = "task_id"), inverseJoinColumns = @JoinColumn(name = "user_id"))
        private Set<User> assignedMembers = new HashSet<>();

        // Pro-Mitglied-Archiv-Status für geteilte Projekt-Tasks: ein zugewiesenes
        // Mitglied kann den Task "für sich" archivieren, ohne ihn für die
        // anderen zu schliessen. Sobald alle zugewiesenen Mitglieder enthalten
        // sind, wird isArchived automatisch global gesetzt (siehe
        // TaskService#applyArchivedChange).
        @Builder.Default
        @ManyToMany(fetch = FetchType.LAZY)
        @JoinTable(name = "task_archived_by", joinColumns = @JoinColumn(name = "task_id"), inverseJoinColumns = @JoinColumn(name = "user_id"))
        private Set<User> archivedBy = new HashSet<>();

        // Ersteller des Tasks - bei persönlichen Tasks identisch zu owner, bei
        // Projekt-Tasks der einzige, der ihn (nach dem Archivieren) für ALLE
        // Mitglieder endgültig löschen darf (siehe TaskService#deleteTask).
        // Bestandstasks von vor diesem Feld haben hier null - dort greift ein
        // Fallback auf den Gruppen-Admin (TaskService#isTaskCreator). Rein
        // serverseitig relevant, daher @JsonIgnore statt über maskMembers
        // anonymisiert zu werden wie assignedMembers/archivedBy.
        @JsonIgnore
        @ManyToOne(fetch = FetchType.LAZY)
        @JoinColumn(name = "created_by_id")
        private User createdBy;

        // Pro-Mitglied-Ausblenden für Projekt-Tasks: ein Nicht-Ersteller kann
        // einen archivierten Gruppen-Task im Archiv "löschen", ohne ihn für die
        // anderen zugewiesenen Mitglieder zu entfernen (TaskService#deleteTask).
        // Rückgängig machbar über PATCH ?hidden=false (Undo-Toast im Frontend).
        @JsonIgnore
        @Builder.Default
        @ManyToMany(fetch = FetchType.LAZY)
        @JoinTable(name = "task_hidden_for", joinColumns = @JoinColumn(name = "task_id"), inverseJoinColumns = @JoinColumn(name = "user_id"))
        private Set<User> hiddenFor = new HashSet<>();

        // Nur für den Cascade beim Löschen (#180): notification_log und
        // task_reminder_override referenzieren task_id per FK (nullable=false),
        // sonst scheitert jeder Delete-Pfad (deleteById, deleteByOwner,
        // deleteByProjectIn) an der Constraint, sobald z.B. eine
        // Überfällig-Benachrichtigung geloggt wurde.
        @JsonIgnore
        @Builder.Default
        @OneToMany(mappedBy = "task", cascade = CascadeType.REMOVE)
        private List<NotificationLog> notificationLogs = new ArrayList<>();

        @JsonIgnore
        @Builder.Default
        @OneToMany(mappedBy = "task", cascade = CascadeType.REMOVE)
        private List<TaskReminderOverride> reminderOverrides = new ArrayList<>();
}

File: frontend/src/types/task.ts
import type { ReminderLeadTime } from "@/types/notifications";

export const TaskFrequency = {
  DAILY: "DAILY",
  WEEKLY: "WEEKLY",
  MONTHLY: "MONTHLY",
  YEARLY: "YEARLY",
  ONCE: "ONCE",
} as const;

export type TaskFrequency = (typeof TaskFrequency)[keyof typeof TaskFrequency];

export const TaskCategory = {
  WORK: "WORK",
  PERSONAL: "PERSONAL",
  SCHOOL: "SCHOOL",
  OTHER: "OTHER",
} as const;

export type TaskCategory = (typeof TaskCategory)[keyof typeof TaskCategory];

/** Wer ein geteiltes Projekt-Task erledigt/zugeordnet hat - Kurzform von User. */
export interface TaskPerson {
  id: string;
  firstName: string;
  lastName: string;
  avatarUrl: string | null;
}

/** Projekt-Referenz auf einem geteilten Gruppen-Task. */
export interface TaskProject {
  id: string;
  name: string;
  isArchived: boolean;
}

export interface Task {
  id: string;
  name: string;
  category: TaskCategory;
  frequency: TaskFrequency;
  description: string;
  dateUntil: string;
  progress: number;
  dateCreated: string;
  /** Vom Server bei jedem Save neu gesetzt - Basis für den Sync-Merge (Last-Write-Wins). */
  updatedAt?: string | null;
  daysInSpan?: number | null;
  amountDid?: number | null;
  /** Zeitpunkt des letzten Fortschritts-Increments - vom Server aus completions abgeleitet (#152). */
  lastAmountDidAt?: string | null;
  /** Completion-Historie als Datums-Strings (YYYY-MM-DD) - Source of Truth für amountDid/progress, Basis für die Verlaufs-Liste und den Abhaken-Toggle (#152). */
  completions?: string[] | null;
  isFavorite?: boolean | null;
  isArchived: boolean;
  durationMinutes?: number | null;
  startTime?: string | null;
  /**
   * Erinnerungs-Vorlauf-Override des aktuellen Users für diesen Task
   * (#102-Follow-up); null/undefined nutzt die Kontoeinstellung. Wird vom
   * Server pro Betrachter befüllt (siehe TaskService#maskMembers) und über
   * einen eigenen Endpoint gesetzt (setTaskReminderLeadTime), nicht über
   * createTask/patchTask.
   */
  reminderLeadTime?: ReminderLeadTime | null;
  /** Gesetzt <=> geteiltes Item eines Gruppen-Projekts statt persönlicher Task. */
  project?: TaskProject | null;
  /** Wer den Task zuletzt als erledigt markiert hat (nur bei Projekt-Tasks relevant). */
  completedBy?: TaskPerson | null;
  /** Self-Service zugewiesene Mitglieder (nur bei Projekt-Tasks relevant). */
  assignedMembers?: TaskPerson[];
  /** Wer den Task bereits individuell für sich archiviert hat (nur bei Projekt-Tasks relevant). */
  archivedBy?: TaskPerson[];
}

/** Fields the server owns and the client must never send on create/patch. */
export type ServerOwnedFields =
  | "id"
  | "dateCreated"
  | "updatedAt"
  | "lastAmountDidAt"
  | "completions"
  | "project"
  | "completedBy"
  | "isArchived"
  | "assignedMembers"
  | "archivedBy"
  | "reminderLeadTime";

/** Payload shape for creating a new task (no id/dateCreated yet). */
export type NewTask = Omit<Task, ServerOwnedFields> & { projectId?: string | null };

/** Options for patchTask: partial task fields plus query-param flags. */
export type PatchTaskOptions = {
  task?: Partial<Omit<Task, ServerOwnedFields>> & { projectId?: string | null };
  resetProgress?: boolean;
  favorite?: boolean;
  archived?: boolean;
  /** Nur noch für geteilte Projekt-Tasks (#152: persönliche Tasks laufen über addTaskCompletion/removeTaskCompletion). */
  amountDid?: number;
  /** Setzt den Task explizit zurück auf persönlich (kein Projekt mehr). */
  unassignProject?: boolean;
  /** Undo für deleteTask bei einem für dich ausgeblendeten Projekt-Task - macht nur mit `false` Sinn, siehe deleteTask. */
  hidden?: boolean;
};
